olm/operator-registry.md

```diff
@@ -0,0 +1,291 @@
+3. Have a way to optimize the build workflow of operator-registry databases (which currently drive OLM's workflows) from previous versions of the database plus new content (ex. new operator at new version) so that the database need not be built from scratch every time they are created.
+
+To start, let's define what the operator-registry does today. It:
+
+1. Takes a set of manifest files and outputs a database file.
+2. Takes a database file and serves a grpc api that serves content from that db
+3. Takes a configmap reference, builds a db and serves a grpc api that serves content from that db
+4. Takes an app-registry path/namespace reference, builds a db and serves a grpc api that serves content from that db
+
+Below are a set of implementation steps that will add new features to accomplish our goals:
+
+*Update the Operator Registry to insert incremental updates*
+
+Add create db, delete, and batch insert APIs to the model layer of operator-registry.
+
+Add a new set of operator registry commands to utilize those new APIs:
+
+`operator-registry create`
+    - inputs: none
+    - outputs: empty operator registry database
+    
+`operator-registry add`
+    - inputs: $operatorBundleImagePath, $dbFile
+    - outputs: updated database file
+    ex: `operator-registry add quay.io/community-operators/foo:sha256@abcd123 example.db`
+    
+`operator-registry add --batch`
+    - inputs: $operatorBundleImagePath, $dbFile
+    - outputs: updated database file
+    ex: `operator-registry add "quay.io/community-operators/foo:sha256@abcd123,quay.io/community-operators/bar:sha256@defg456" example.db`
+    
+`operator-registry delete`
+    - inputs: $operatorName, $dbFile
+    - outputs: updated database file without $operatorName included
+    ex: `operator-registry delete bar example.db`
+
+`operator-registry delete-latest`
+    - inputs: $operatorName, $dbFile
+    - outputs: updated database file without latest version of $operatorName included
+    ex: `operator-registry delete-latest foo example.db`
+    
+*Reference non latest versions of bundles by image digests*
+
+Currently the operator-registry database has a table to the effect of:
+
+OperatorBundle
+    -name: foo
+    -bundle: "{$jsonblob}"
+    
+We will add a field to this table that will reference the bundle image path and, when the operator is not the latest version of the default channel, will not include the bundle blob itself:
+
+OperatorBundle
+    -name: foo
+    -bundle: NULL
+    -bundlePath: "quay.io/community-operators/foo:sha256@abcd123"
+    
+Additionally, we will update the grpc API that OLM currently uses to get the bundle that if it attempts to pull the non latest version will return null. In that case, it will make a second query to get the bundlePath. Then we will create a pod with that image that will serve the bundle data by writing the data to a configmap that can be read by OLM.
```

Comment:
Is there an unstated assumption here that the index will contain the complete set of metadata+manifests for the latest version of all operators? And that it will then only contain metadata for all non-latest versions of the operator?

Reply:
It is actually explicitly called out in this doc, it's just not explicitly called out in the implementation detail section about the database add method. Instead it's called out explicitly in the index container definition user story a little further up this document.

If it would make it more clear, I can update the registry api section of the implementation details to call that part out.